### ISSUE TITLE
always rotate introduction points per failback instance (don't repeat them if we have options to choose from)

### ISSUE THREAD

**Issue by gits7r** [DonnchaC/onionbalance#18]
This will allow us to use OnionBalance in cascade (onionbalanced failback instances) as well as load balance the introduction points per failback instance, which is important. This should be applied when context allows us, regardless if running in 'basic mode' or 'complex mode'.

Let's assume we have a master OnionBalance hidden service: master.onion running in basic mode. It has 3 failback instances a.onion; b.onion; c.onion. It will select 3 introduction points per failback instance (total 9 introduction points for master.onion). OnionBalance publishes a new master descriptor every hour, or earlier if one (or more) of the failback instances change their introduction points for any reason.

OnionBalance should remember the introduction points for each failback instance published in the last master descriptor and avoid repeating them if possible. Choose random all the time, but ROTATE introduction points per failback instance in the master descriptor.

Example:
a.onion has a_ip1; a_ip2; a_ip3; a_ip4; a_ip5; a_ip6; a_ip7.
b.onion has b_ip1; b_ip2; b_ip3; b_ip4; b_ip5; b_ip6; b_ip7.
c.onion has c_ip1; c_ip2; c_ip3; c_ip4; c_ip5; c_ip6.

master.onion randomly selects in descriptor_1 9 introduction points as follows:
a_ip2; a_ip6; a_ip5; b_ip1; b_ip7; b_ip4; c_ip6; c_ip2; c_ip3

If for any reason we need to republish the master descriptor (one hour passed or a failback instance changed introduction point(s), etc), when we publish descriptor_2, we remember descriptor_1's introduction points for each failback instance and we never repeat them if we have other options to choose from, but we choose again totally random from the remaining valid options:
descriptor_2 9 introduction points:
a_ip1; a_ip3; a_ip7; b_ip2; b_ip3; b_ip5; c_ip5; c_ip1; c_ip4

If we don't have enough options to choose from (if for example we have 3 failback instances and all have 3 introduction points each, then we must select all of them all the time, so we don't have the choice to repeat or not repeat them regardless if they change or not) we should choose same IPs again - no other option here.

If we can only select one new IP from each failback instance, so be it.
If we can only select one new IP in total (from all failback instances combined), so be it. 

Basically we remember, randomize and avoid repeating IPs for each failback instance when we have a pool of IPs per failback instance from where to choose from. If we don't have a pool big enough to have the all the IPs new, use as many new as possible, if not possible at all, so be it (republish the same master descriptor with new time stamp).

**Comment by DonnchaC** [comment]
The distinct descriptors feature (implemented in 54444b0) improves the process for selecting introduction points. I'm closing this for now as OnionBalance can now use up to 60 introduction points without "cascading".

@gits7r Please reopen this ticket if you think the distinct descriptor feature doesn't solve the spirit of this issue. Thanks again for all your contributions!
